avatars/avatars.js

```diff
@@ -1883,7 +1891,72 @@ class Avatar {
       }
       this.debugMesh.visible = debug.enabled;
     }
-	}
+
+    //#################################### manually set mouth movement ##########################################
+    const _handleMouthMovementAttack=()=>{
+      this.volume = ((timestamp/1000 - this.manuallySetMouthStartTime) / this.manuallySetMouthAttackTime)/12;
+      if(timestamp/1000 - this.manuallySetMouthStartTime >= this.manuallySetMouthAttackTime){
+        this.manuallySetMouthState = 'decay';
+        this.manuallySetMouthStartTime = timestamp/1000;
+      }
+    }
+    const _handleMouthMovementDecay=()=>{
+      this.volume = (1 - ((timestamp/1000 - this.manuallySetMouthStartTime) / this.manuallySetMouthDecayTime) * 0.8)/12;
+      if(timestamp/1000 - this.manuallySetMouthStartTime >= this.manuallySetMouthDecayTime){
+        this.manuallySetMouthState='sustain';
+        this.manuallySetMouthStartTime = timestamp/1000;
+      }
+    }
+    const _handleMouthMovementSustain=()=>{
+      if(timestamp/1000 - this.manuallySetMouthStartTime >= this.manuallySetMouthSustainTime){
+        this.manuallySetMouthState='release';
+        this.manuallySetMouthStartTime = timestamp/1000;
+      } 
+    }
+    const _handleMouthMovementRelease=()=>{
+      this.volume = (0.2 - ((timestamp/1000 - this.manuallySetMouthStartTime) / this.manuallySetMouthReleaseTime) * 0.2)/12;
+      if(timestamp/1000 - this.manuallySetMouthStartTime >= this.manuallySetMouthReleaseTime){
+        this.manuallySetMouthState=null;
+        this.manuallySetMouth=false;
+        this.manuallySetMouthStartTime = -1;
+      }
+    }
+    const _handleMouthMovementNull=()=>{
+      this.manuallySetMouthState = this.manuallySetMouth ? 'attack' : null;
+      this.manuallySetMouthStartTime = timestamp/1000;
+    }
+    switch (this.manuallySetMouthState) {
+      case 'attack': {
+        _handleMouthMovementAttack();
+        break;
+      }
+      case 'decay': {
+        _handleMouthMovementDecay();
+        break;
+      }
+      case 'sustain': {
+        _handleMouthMovementSustain();
+        break;
+      }
+      case 'release': {
+        _handleMouthMovementRelease();
+        break;
+      }
+      case null: {
+        _handleMouthMovementNull();
+        break;
+      }
+    }
+
+  }
+  setMouthMoving(attack, decay, sustain, release){
```

Comment:
This should be abstracted better. The avatars implementation is already too complicated, especially since we already have a volume tracking implementation and this is kind of a second one.

I would say that maybe this belongs in `CharacterSfx` since that is most related but since this does not actually generate audio, maybe it belongs in a new class like `CharacterBehavior`?

Reply:
Then, `CharacterBehavior` could calculate and set the volume on the avatar at player -> avatar binding update time.